src/Npgsql/Internal/ResolverFactories/JsonDynamicTypeInfoResolverFactory.cs

```diff
@@ -116,9 +116,14 @@ void AddUserMappings(bool jsonb, Type[] clrTypes)
             {
                 var jsonb = dataTypeName == DataTypeNames.Jsonb;
 
-                // For jsonb we can't properly support polymorphic serialization unless we do quite some additional work
-                // so we default to mapping.Type instead (exact types will never serialize their "$type" fields, essentially disabling the feature).
+                // For jsonb we can't properly support polymorphic serialization unless the SerializerOptions.AllowOutOfOrderMetadataProperties is `true`.
+                // If `jsonb` AND `AllowOutOfOrderMetadataProperties` is `false`, use `mapping.Type` as the base type for the converter, and 
+                // strip the "$type" field; essentially disabling the feature.
+                #if NET9_0_OR_GREATER
```

Comment:
Looks like it works, but the assertions will have to change due to the property order changing.

Reply:
Ok, updated; that suite seems to pass now

![image](https://github.com/user-attachments/assets/a9c106c0-a6e8-4db0-84e9-26265e471460)

Reply:
> I was also wondering if you thought that 'out of order' should be the default behavior.. I don't think it would hurt performance of serialization, and property order would be preserved for deserializing JSON (as opposed to JSONB), so IMO it would be win-win to just support it rather than runtime erroring.

AllowOutOfOrderMetadataProperties isn't the default in S.T.Json itself, and there are good reasons for that - this requires buffering inside S.T.Json, which could lead to increased memory usage etc. (https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-9/); if we enabled this by default, people not needing JSON polymorphism here would be paying that for no reason. On the other hand, it's true that this is a place where we **know** that the properties wouldn't be preserved (PG jsonb), so there's definitely more justification to enable it by default.

I'm still leaning towards requiring the user to explicitly opt in, and for Npgsql to simply not mess with the JsonSerializationOptions. @NinoFloris @vonzshik what do you think?

Reply:
IMO I agree that it should be an explicit opt in. It's not enabled by default in STJ and I trust them a bit more rather than us assuming that "it's definitely not gonna blow into our face".

Reply:
> I'd rather we bump our STJ package version so we can access this new property downlevel too.

@NinoFloris just noting that STJ 9.0 will be out of support a bit before .NET 8.0; so people doing LTS through and through may not want to take a reference on a non-LTS package.

So it might be better to just keep this functionality for people doing .NET 9.0 (with the logic that if you're OK doing STJ 9.0, you should be OK doing .NET 9.0 as well). But I don't have strong feelings on this, so I'll defer to whatever you want.

Reply:
It's a reasonable thing to consider yeah. Though by and large we treat STJ as a normal package, meaning it lives outside 'inbox version' SDK LTS cycles. 

If that nebulous group of .NET 8.0 users doing LTS through and through approach it from such a theoretical standpoint that they feel none of their OSS deps are allowed to version upwards either then maybe they shouldn't depend on Npgsql 9.0. It's likely many of those users also run EF where 8.0 would not even be compatible with Npgsql 9.0. Would it be ok for those users to depend on EF 9.0, it being non-LTS? I don't think we should try too hard to keep our 'STS' versions entirely version aligned to LTS dependencies (not that I'd advocate for breaking if we can avoid it either). 

@vonzshik do you have any thoughts on this?

Reply:
>I'm still leaning towards requiring the user to explicitly opt in, and for Npgsql to simply not mess with the JsonSerializationOptions. @NinoFloris @vonzshik what do you think?

Yes I think so too. 

Though we could consider the following case: when `AllowOutOfOrderMetadataProperties` is false, and a type is found in jsonbClrTypes that has any `PolymorphismOptions.DerivedTypes` it would help if we could surface to the user somehow that polymorphism won't work. It might be worth a log entry. Throwing would be problematic as it would force users to opt into `AllowOutOfOrderMetadataProperties` even though they may not want to use polymorphism in the context of Npgsql at all.

Reply:
> It might be worth a log entry

Yeah, that would make sense to me...